### PR TITLE
Add several multi_queue_executor optimizations, automatic work distribution across streams, hipSYCL_prefer_execution_lane extension

### PR DESCRIPTION
**Pull request by illuhad**
Adds multiple features to DAG processing and `multi_queue_executor`:

1. This adds two important optimizations to DAG processing and the `multi_queue_executor`:
* Add the optimization to simplify the DAG early on by removing requirements that are already implied by other requirements because they already appear in their requirement graph. This has been on the todo list for a long time and helps avoid massive graph growth when huge number of tasks are submitted as discussed in #380.
* Add the `multi_queue_executor` optimization that synchronization with other execution lanes/streams is only necessary for the last dependency on a lane because of the in-order execution property of lanes/streams. In other cases, the synchronization can be removed.

2. Additionally, this finally adds **initial support for automatic work distribution across the lane/stream pool of a device**, allowing our runtime architecture to finally flex its muscles. Currently, the distribution is done by first attempting to minimize the number of dependencies on other lanes. This effectively causes task groups with dependencies between them to tend to execute on the same lane/stream, reducing latency. When this criteria is inconclusive, the `multi_queue_executor` will submit the task to a lane/stream that it thinks is underutilized. For this purpose, it records the lanes to which the last 100 tasks were submitted. A time-decay mechanism is added such that older information contributes less.
This approach is still a bit simplistic in some aspects, but it's a start.

3. Adds the `hipSYCL_prefer_execution_lane` extension. This is a command group property that carries a hint for the `multi_queue_executor` on which backend lane/stream to execute the operation:
```
std::size_t lane_id = 0; // Execute on first lane
q.submit({property::command_group::hipSYCL_prefer_execution_lane{lane_id}},
  [&](cl::sycl::handler &cgh) {
   cgh.parallel_for(...);
});
```
The operation will be executed on the provided lane id modulo the number of lanes.
This can be used in circumstances where the `multi_queue_executor` does not schedule in an optimal fashion.

- [x] TODO: We should expose the configuration used internally using the `settings` API to the user instead of just hardcoding numbers.

Now adds the following configuration parameters via the `settings` infrastructure:
* `HIPSYCL_RT_DAG_REQ_OPTIMIZATION_DEPTH`: maximum depth when descending the DAG requirement tree to look for DAG optimization opportunities, such as eliding unnecessary dependencies.
* `HIPSYCL_RT_MQE_LANE_STATISTICS_MAX_SIZE`: For the `multi_queue_executor`, the maximum size of entries in the lane statistics, i.e. the maximum number of submissions to retain statistical information about. This information is used to estimate execution lane utilization.
* `HIPSYCL_RT_MQE_LANE_STATISTICS_DECAY_TIME`: The time in seconds (floating point value) after which to forget information about old submissions.

### CODE REVIEW

**Comment by illuhad** [comment]
@hipSYCL-ci-bot test again.

File: include/hipSYCL/runtime/data.hpp [hunk 1]
```diff
@@ -241,8 +241,13 @@ class data_user_tracker
...
   void for_each_user(F f){
     std::lock_guard<std::mutex> lock{_lock};
-    for(auto& user : _users) {
-      f(user);
+    // Iterate in reverse order over the users since
+    // this will iterate over the newest users first.
+    // This is a more advantageous pattern e.g. during
+    // DAG construction as it allows finding the relevant users
+    // quicker.
+    for(int i = _users.size() - 1; i >= 0; --i) {
```
Comment (fodinabor) [inline include/hipSYCL/runtime/data.hpp:249]: not sure this is actually more useful..?
There's no early break or something, so we don't get a real advantage here by using a less uniform access pattern?

File: include/hipSYCL/runtime/multi_queue_executor.hpp [hunk 2]
```diff
@@ -40,6 +41,81 @@
...
+    std::size_t count = 0;
+    for(const auto& s : _last_submissions) {
+      if(s.bin == bin) ++count;
+    }
+    return count;
+  }
+
+  template<class WeightFunc>
+  std::vector<double> build_weighted_bins(WeightFunc w) const {
+    std::vector<double> bins_out = std::vector<double>(_num_bins);
```
Comment (fodinabor) [inline include/hipSYCL/runtime/multi_queue_executor.hpp]:
```suggestion
    std::vector<double> bins_out(_num_bins);
```

File: include/hipSYCL/runtime/multi_queue_executor.hpp [hunk 3]
```diff
@@ -40,6 +41,81 @@
...
+    for(const auto& s : _last_submissions) {
+      bins_out[s.bin] += w(s.timestamp);
+    }
+    return bins_out;
+  }
+
+  std::vector<double>
+  build_decaying_bins() const {
+    std::size_t now = this->now();
+    return build_weighted_bins([&,this](std::size_t timestamp) -> double {
```
Comment (fodinabor) [inline include/hipSYCL/runtime/multi_queue_executor.hpp]: I don't see a reason for capturing by reference..? if I see this correctly, just capture `now` as copy explicitly and you should be fine (and have an indirection less :))

File: src/runtime/dag_node.cpp [hunk 4]
```diff
@@ -128,13 +145,67 @@ const execution_hints &dag_node::get_execution_hints() const { return _hints; }
...
+
 // Add requirement if not already present
 void dag_node::add_requirement(dag_node_ptr requirement)
 {
   for (auto req : _requirements) {
     if (req == requirement)
       return;
   }
+
+  auto is_reachable_from = [&, this](dag_node_ptr from, dag_node_ptr to,
```
Comment (fodinabor) [inline src/runtime/dag_node.cpp]: no need for capturing anything.. don't really get the use of the lambda anyways.. just renaming `recursive_find`? :D

**Review by fodinabor — COMMENTED** [review]
Any benchmarks yet on reduced / increased scheduling overhead? :)

File: include/hipSYCL/runtime/data.hpp (hunk 1 above)
Comment (illuhad) [reply include/hipSYCL/runtime/data.hpp:249]: Yes, there is an early exit, but it's not necessarily obvious: https://github.com/illuhad/hipSYCL/blob/6282d829d20ff0a2e71c39a707e8627c8da858c3/src/runtime/dag_node.cpp#L194

**Comment by illuhad** [comment]
@fodinabor no benchmarks yet, but I don't expect there to be a lot of additional overhead. *Maybe* the early DAG optimization if you have DAGs with a lot of requirements, but in that case you can just reduce `HIPSYCL_RT_DAG_REQ_OPTIMIZATION_DEPTH` to prevent it from descending too far down the requirement tree. The MQE stuff is probably only noticable because of likely substantial performance improvements for small kernels because they can a) run in parallel and b) fewer `cuda/hipStreamWaitEvent` are submitted.

File: src/runtime/dag_node.cpp (hunk 4 above)
Comment (illuhad) [reply src/runtime/dag_node.cpp]: The captures are remnants of a previous design ;) I decided to keep the lambda as I think it makes somewhat clearer what happens and the `recursive_find` function has the right name imo for what it does.

**Pull request merged**